cpp/src/utilities/cudf_utils.h

```diff
@@ -25,9 +25,49 @@ bool gdf_is_valid(const gdf_valid_type *valid, gdf_index_type pos) {
 		return true;
 }
 
+/**
+  * Calculates the number of chars used for a validity indicator pseudo-column for a given column's size.
+  *
+  * @note Note that this function assumes that `gdf_valid_type` is unsigned char
+  * @node This function is different gdf_get_num_bytes_for_valids_allocation because it refers to bytes used as opposed to allocated
+  *
+  * @param[in] column_size the number of elements
+  * @return the number of bytes necessary to make available for the validity indicator pseudo-column
+  */
 CUDA_HOST_DEVICE_CALLABLE
-gdf_size_type gdf_get_num_chars_bitmask(gdf_size_type size) { 
-	return (( size + ( GDF_VALID_BITSIZE - 1)) / GDF_VALID_BITSIZE ); 
+gdf_size_type gdf_get_num_chars_bitmask(gdf_size_type column_size) { 
+	return (( column_size + ( GDF_VALID_BITSIZE - 1)) / GDF_VALID_BITSIZE ); 
+}
+
+// Buffers are padded to 64-byte boundaries (for SIMD) static
+constexpr int32_t kArrowAlignment = 64;
+
+// Tensors are padded to 64-byte boundaries static
+constexpr int32_t kTensorAlignment = 64;
+
+// Align on 8-byte boundaries in IPC static 
+constexpr int32_t kArrowIpcAlignment = 8;
+
+// Align on 4-byte boundaries in CUDF static 
+constexpr int32_t kCudfIpcAlignment = 4;
+
+ //todo, enable arrow ipc utils, and remove this method 
+CUDA_HOST_DEVICE_CALLABLE
+static gdf_size_type PaddedLength(int64_t nbytes, int32_t alignment = kCudfIpcAlignment) {   
+	return ((nbytes + alignment - 1) / alignment) * alignment; 
+}
+
+/**
+  * Calculates the number of bytes to allocate for a validity indicator pseudo-column for a given column's size.
+  *
+  * @note Note that this function assumes the valids need to be allocated to be aligned with a 4 byte boundary
+  *
+  * @param[in] column_size the number of elements
+  * @return the number of bytes necessary to allocate for the validity indicator pseudo-column
+  */
+CUDA_HOST_DEVICE_CALLABLE
+gdf_size_type gdf_get_num_bytes_for_valids_allocation(gdf_size_type column_size) { 
```

Comment:
Please just use `gdf_get_num_chars_bitmask`. 

We've already got a set of centralized bitmask utilities here: https://github.com/rapidsai/cudf/blob/branch-0.6/cpp/src/bitmask/bit_mask.h

They aren't used yet because replacing `gdf_valid_type` with `bit_mask_t` can't be done in pieces and will need to be done all at once.

Reply:
This a little different. We want valids to be aligned AND padded to 4 bytes. So allocations should use this function. We found some errors that went away when we started making allocations that were padded to 4 bytes.

Reply:
Yeah, the new utilities will pad to 64B.

You may get some cuda-memcheck errors if it's not padded to 4B, but I've yet to see it actually cause any crashes. Did you see actual program crashes?

Reply:
This change is being done to fix a memory issue. You want me to undo the fix and leave the memory issue? We have had program crashed due to these issues. More are fixed here: https://github.com/rapidsai/cudf/pull/1053